pennylane_lightning/core/src/simulators/lightning_kokkos/catalyst/LightningKokkosObsManager.hpp

```diff
@@ -0,0 +1,210 @@
+// Copyright 2023-2024 Xanadu Quantum Technologies Inc.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <array>
+#include <stdexcept>
+#include <tuple>
+#include <utility>
+
+#include "Types.h"
+#include "Utils.hpp"
+
+#include "ObservablesKokkos.hpp"
+
+namespace Catalyst::Runtime::Simulator {
+
+/**
+ * @brief The LightningKokkosObsManager caches observables of a program at
+ * runtime and maps each one to a const unique index (`int64_t`) in the scope of
+ * the global context manager.
+ */
+template <typename PrecisionT> class LightningKokkosObsManager {
+  private:
+    using VectorStateT =
+        Pennylane::LightningKokkos::StateVectorKokkos<PrecisionT>;
+    using ObservableClassName =
+        Pennylane::Observables::Observable<VectorStateT>;
+    using ObservablePairType =
+        std::pair<std::shared_ptr<ObservableClassName>, ObsType>;
```

Comment:
`ObsType` is defined in the Catalyst header, see here:
https://github.com/PennyLaneAI/catalyst/blob/bcdd27cd313e1893d24492cdcbe4a00bbd114611/runtime/include/Types.h#L37-L50

The observable system in the runtime is based around backends using their own representation for observables. The only thing that is expected is that observables can be "constructed" first, and then used in other runtime calls. 
So when the program is running it might call `__catalyst__qis__NamedObs` to instruct the device to "construct" a Hadamard observable, but what the observable looks like is left to the device and the program only receives a key (of type `ObsIdType`) to refer to this observable later. 
`ObsId` is just enum used to select from a fixed set of known observables, like the paulis.
The program can then build compound observables from other observables, or the observable key can be used in a call to a measurement process.

`ObsType` is another enum to distinguish between three kinds of observables: a basic observable, or a compound observable, like products and sums.

To be honest the names are not great at distinguishing what each of them represents.

Reply:
All good, just had a hard time finding the definition. We can all collectively decide on a hierarchy of names at a later date to unify meaning across the packages.